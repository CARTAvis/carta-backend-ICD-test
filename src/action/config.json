{
    "serverURL": "wss://carta.asiaa.sinica.edu.tw/socketdev",
    "serverURL0": "ws://127.0.0.1:3002",
    "serverURL1": "ws://cartaviewer:5000",
    "path": {
        "root": ".",
        "base": "$BASE",
        "directory": ".",
        "QA": "set_QA",
        "region": "set_QA/set_QA_regionTest",
        "performance": "set_QA_performance"
    },
    "image": {
        "singleChannel": "cube_A/cube_A_01600_z00001.fits",
        "cube": "cube_A/cube_A_00800_z00100.fits"
    },
    "timeout": {
        "connection": 1000,
        "listFile": 1000,
        "openFile": 1000,
        "readFile": 6100,
        "readLargeImage": 60000,
        "changeChannel": 3000,
        "playImages": 8000,
        "cubeHistogram": 500000,
        "cancel": 5000,
        "concurrent": 4000,
        "messageEvent": 500,
        "mouseEvent": 400,
        "region": 500,
        "import": 500,
        "export": 500,
        "resume": 10000,
        "sleep": 3000,
        "playAnimator": 140000,
        "playLargeAnimator": 600000,
        "contour": 4000,
        "zoom": 12200
    },
    "repeat": {
        "concurrent": 10,
        "cursor": 20,
        "image": 10,
        "animation": 10,
        "resumeSession": 10,
        "contour": 10,
        "zoom": 5
    },
    "log": {
        "error": false,
        "warning": true,
        "event": false
    }
}