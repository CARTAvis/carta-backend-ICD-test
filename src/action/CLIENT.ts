import { CARTA } from "carta-protobuf";

import config from "./config.json";
let nodeusage = require("usage");
const fs = require("fs");
const procfs = require("procfs-stats");
export class Client {
    IcdVersion: number = 17;
    CartaType = new Map<number, any>([
        [0, CARTA.ErrorData],
        [1, CARTA.RegisterViewer],
        [2, CARTA.FileListRequest],
        [3, CARTA.FileInfoRequest],
        [4, CARTA.OpenFile],
        [6, CARTA.SetImageChannels],
        [7, CARTA.SetCursor],
        [8, CARTA.SetSpatialRequirements],
        [9, CARTA.SetHistogramRequirements],
        [10, CARTA.SetStatsRequirements],
        [11, CARTA.SetRegion],
        [12, CARTA.RemoveRegion],
        [13, CARTA.CloseFile],
        [14, CARTA.SetSpectralRequirements],
        [15, CARTA.StartAnimation],
        [16, CARTA.StartAnimationAck],
        [17, CARTA.StopAnimation],
        [18, CARTA.RegisterViewerAck],
        [19, CARTA.FileListResponse],
        [20, CARTA.FileInfoResponse],
        [21, CARTA.OpenFileAck],
        [22, CARTA.SetRegionAck],
        [23, CARTA.RegionHistogramData],
        [25, CARTA.SpatialProfileData],
        [26, CARTA.SpectralProfileData],
        [27, CARTA.RegionStatsData],
        [28, CARTA.ErrorData],
        [29, CARTA.AnimationFlowControl],
        [30, CARTA.AddRequiredTiles],
        [31, CARTA.RemoveRequiredTiles],
        [32, CARTA.RasterTileData],
        [33, CARTA.RegionListRequest],
        [34, CARTA.RegionListResponse],
        [35, CARTA.RegionFileInfoRequest],
        [36, CARTA.RegionFileInfoResponse],
        [37, CARTA.ImportRegion],
        [38, CARTA.ImportRegionAck],
        [39, CARTA.ExportRegion],
        [40, CARTA.ExportRegionAck],
        [41, CARTA.SetUserPreferences],
        [42, CARTA.SetUserPreferencesAck],
        [43, CARTA.SetUserLayout],
        [44, CARTA.SetUserLayoutAck],
        [45, CARTA.SetContourParameters],
        [46, CARTA.ContourImageData],
        [47, CARTA.ResumeSession],
        [48, CARTA.ResumeSessionAck],
        [49, CARTA.RasterTileSync],
        [50, CARTA.CatalogListRequest],
        [51, CARTA.CatalogListResponse],
        [52, CARTA.CatalogFileInfoRequest],
        [53, CARTA.CatalogFileInfoResponse],
        [54, CARTA.OpenCatalogFile],
        [55, CARTA.OpenCatalogFileAck],
        [56, CARTA.CloseCatalogFile],
        [57, CARTA.CatalogFilterRequest],
        [58, CARTA.CatalogFilterResponse],
        [59, CARTA.ScriptingRequest],
        [60, CARTA.ScriptingResponse],
        [67, CARTA.SpectralLineRequest],
        [68, CARTA.SpectralLineResponse],

    ]);
    CartaTypeValue(type: any): number {
        let ret: number = 0;
        for (let [key, value] of this.CartaType.entries()) {
            if (value === type)
                ret = key;
        }
        return ret;
    }
    static eventCount = { value: 0 };
    connection: WebSocket;
    // Construct a websocket connection to url
    constructor(url: string) {
        this.connection = new WebSocket(url);
        this.connection.binaryType = "arraybuffer";
    }
    open(timeout?: number) {
        return new Promise<void>((resolve, reject) => {
            this.connection.onopen = onOpen;
            function onOpen(this: WebSocket, ev: Event) {
                resolve();
            }
            if (timeout) {
                let failTimer = setTimeout(() => {
                    clearTimeout(failTimer);
                    reject();
                }, timeout);
            }
        });
    }
    close(timeout?: number) {
        this.connection.close();
        return new Promise<void>((resolve, reject) => {
            this.connection.onclose = onClose;
            function onClose(this: WebSocket, ev: Event) {
                resolve();
            }
            if (timeout) {
                let failTimer = setTimeout(() => {
                    clearTimeout(failTimer);
                    reject();
                }, timeout);
            }
        });
    }
    /// Send websocket message async
    /// Parameters: connection(Websocket ref), cartaType(CARTA.type), eventMessage(the sending message)
    /// return a Promise<any> for await
    send(cartaType: any, eventMessage: any, ) {
        return new Promise<void>(resolve => {
            let message = cartaType.create(eventMessage);
            let payload = cartaType.encode(message).finish();
            let eventData = new Uint8Array(8 + payload.byteLength);
            const eventHeader16 = new Uint16Array(eventData.buffer, 0, 2);
            const eventHeader32 = new Uint32Array(eventData.buffer, 4, 1);
            eventHeader16[0] = this.CartaTypeValue(cartaType);
            eventHeader16[1] = this.IcdVersion;
            eventHeader32[0] = Client.eventCount.value++; // eventCounter;
            if (config.log.event) {
                console.log(`${cartaType.name} => @ ${eventHeader32[0]}`);
            }

            eventData.set(payload, 8);
            this.connection.send(eventData);

            resolve();
        });
    }
    /// Receive websocket message async
    /// Parameters: connection(Websocket ref), cartaType(CARTA.type)
    /// timeout: promise will return nothing until time out if timeout > 0
    /// return a Promise<cartaType> for await
    receive(cartaType: any, timeout?: number, isReceive?: boolean) {
        return new Promise<any>((resolve, reject) => {
            this.connection.onmessage = async (messageEvent: MessageEvent) => {
                const eventHeader16 = new Uint16Array(messageEvent.data, 0, 2);
                const eventHeader32 = new Uint32Array(messageEvent.data, 4, 1);
                const eventData = new Uint8Array(messageEvent.data, 8);

                const eventNumber = eventHeader16[0];
                const eventIcdVersion = eventHeader16[1];
                const eventId = eventHeader32[0];
                if (config.log.event) {
                    console.log(`<= ${this.CartaType.get(eventNumber).name} @ ${eventId}`);
                }

                if (eventIcdVersion !== this.IcdVersion && config.log.warning) {
                    console.warn(`Server event has ICD version ${eventIcdVersion}, which differs from frontend version ${this.IcdVersion}. Errors may occur`);
                }

                if (this.CartaType.get(eventNumber) === cartaType) {
                    if (timeout && !isReceive) {
                        reject();
                    }
                    let data;
                    switch (cartaType) {
                        case CARTA.SpatialProfileData:
                            data = CARTA.SpatialProfileData.decode(eventData);
                            data.profiles = data.profiles.map(p => processSpatialProfile(p));
                            break;
                        case CARTA.SpectralProfileData:
                            data = CARTA.SpectralProfileData.decode(eventData);
                            data.profiles = data.profiles.map(p => processSpectralProfile(p));
                            break;
                        default:
                            data = cartaType.decode(eventData);
                            break;
                    }
                    resolve(data);
                }
            };
            if (timeout) {
                let Timer = setTimeout(() => {
                    clearTimeout(Timer);
                    if (isReceive) {
                        reject();
                    } else {
                        resolve();
                    }
                }, timeout);
            }
        });
    }
    /// A receiving websocket message in any type async
    /// timeout: promise will return CARTA data until time out if timeout > 0
    /// return a Promise<any> for await
    receiveAny(timeout?: number) {
        return new Promise<any>((resolve, reject) => {
            this.connection.onmessage = async (messageEvent: MessageEvent) => {
                const eventHeader16 = new Uint16Array(messageEvent.data, 0, 2);
                const eventHeader32 = new Uint32Array(messageEvent.data, 4, 1);
                const eventData = new Uint8Array(messageEvent.data, 8);
                const eventNumber = eventHeader16[0];
                // const eventIcdVersion = eventHeader16[1];
                const eventId = eventHeader32[0];
                if (config.log.event) {
                    console.log(`<= ${this.CartaType.get(eventNumber).name} @ ${eventId}`);
                }

                let data;
                let type = this.CartaType.get(eventNumber);
                switch (type) {
                    case CARTA.SpatialProfileData:
                        data = CARTA.SpatialProfileData.decode(eventData);
                        data.profiles = data.profiles.map(p => processSpatialProfile(p));
                        break;
                    case CARTA.SpectralProfileData:
                        data = CARTA.SpectralProfileData.decode(eventData);
                        data.profiles = data.profiles.map(p => processSpectralProfile(p));
                        break;
                    default:
                        data = type.decode(eventData);
                        break;
                }
                resolve(data);
            };
            if (timeout) {
                let Timer = setTimeout(() => {
                    clearTimeout(Timer);
                    reject();
                }, timeout);
            }
        });
    }
    /// Receive CARTA stream async
    /// Until the number: totalCount of mesaages have received
    stream(count?: number) {
        if (count <= 0) {
            return Promise.resolve();
        }

        let _count: number = 0;
        let ack: AckStream = {
            Responce: [],
            RasterTileData: [],
            RasterTileSync: [],
            SpatialProfileData: [],
            RegionStatsData: [],
            RegionHistogramData: [],
            SpectralProfileData: [],
            ContourImageData: [],
            CatalogFilterResponse: [],
        };

        return new Promise<AckStream>(resolve => {
            this.connection.onmessage = (messageEvent: MessageEvent) => {
                const eventHeader16 = new Uint16Array(messageEvent.data, 0, 2);
                const eventHeader32 = new Uint32Array(messageEvent.data, 4, 1);
                const eventData = new Uint8Array(messageEvent.data, 8);

                const eventNumber = eventHeader16[0];
                const eventIcdVersion = eventHeader16[1];
                const eventId = eventHeader32[0];

                if (eventIcdVersion !== this.IcdVersion && config.log.warning) {
                    console.warn(`Server event has ICD version ${eventIcdVersion}, which differs from frontend version ${this.IcdVersion}. Errors may occur`);
                }
                if (config.log.event) {
                    console.log(`<= ${this.CartaType.get(eventNumber).name} @ ${eventId}`);
                }

                let data;
                switch (this.CartaType.get(eventNumber)) {
                    default:
                        ack.Responce.push(this.CartaType.get(eventNumber).decode(eventData));
                        break;
                    case CARTA.RasterTileData:
                        ack.RasterTileData.push(CARTA.RasterTileData.decode(eventData));
                        break;
                    case CARTA.RasterTileSync:
                        ack.RasterTileSync.push(CARTA.RasterTileSync.decode(eventData));
                        break;
                    case CARTA.RegionStatsData:
                        ack.RegionStatsData.push(CARTA.RegionStatsData.decode(eventData));
                        break;
                    case CARTA.RegionHistogramData:
                        ack.RegionHistogramData.push(CARTA.RegionHistogramData.decode(eventData));
                        break;
                    case CARTA.SpatialProfileData:
                        data = CARTA.SpatialProfileData.decode(eventData);
                        data.profiles = data.profiles.map(p => processSpatialProfile(p));
                        ack.SpatialProfileData.push(data);
                        break;
                    case CARTA.SpectralProfileData:
                        data = CARTA.SpectralProfileData.decode(eventData);
                        data.profiles = data.profiles.map(p => processSpectralProfile(p));
                        ack.SpectralProfileData.push(data);
                        break;
                    case CARTA.ContourImageData:
                        ack.ContourImageData.push(CARTA.ContourImageData.decode(eventData));
                        break;
                    case CARTA.CatalogFilterResponse:
                        ack.CatalogFilterResponse.push(CARTA.CatalogFilterResponse.decode(eventData));
                        break;
                }

                _count++;
                if (_count === count) {
                    resolve(ack);
                }
            };
        });
    }
};

export interface AckStream {
    Responce: any[],
    RasterTileData: CARTA.RasterTileData[];
    RasterTileSync: CARTA.RasterTileSync[];
    SpatialProfileData: CARTA.SpatialProfile[];
    RegionStatsData: CARTA.RegionStatsData[];
    RegionHistogramData: CARTA.RegionHistogramData[];
    SpectralProfileData: CARTA.SpectralProfileData[];
    ContourImageData: CARTA.ContourImageData[];
    CatalogFilterResponse: CARTA.CatalogFilterResponse[];
}
interface ProcessedSpatialProfile extends CARTA.ISpatialProfile { values: Float32Array; }
function processSpatialProfile(profile: CARTA.ISpatialProfile): ProcessedSpatialProfile {
    if (profile.rawValuesFp32 && profile.rawValuesFp32.length && profile.rawValuesFp32.length % 4 === 0) {
        return {
            coordinate: profile.coordinate,
            start: profile.start,
            end: profile.end,
            values: new Float32Array(profile.rawValuesFp32.slice().buffer)
        };
    }
    return {
        coordinate: profile.coordinate,
        start: profile.start,
        end: profile.end,
        values: null
    };
}
interface ProcessedSpectralProfile extends CARTA.ISpectralProfile { values: Float32Array | Float64Array; }
function processSpectralProfile(profile: CARTA.ISpectralProfile): ProcessedSpectralProfile {
    if (profile.rawValuesFp64 && profile.rawValuesFp64.length && profile.rawValuesFp64.length % 8 === 0) {
        return {
            coordinate: profile.coordinate,
            statsType: profile.statsType,
            values: new Float64Array(profile.rawValuesFp64.slice().buffer)
        };
    } else if (profile.rawValuesFp32 && profile.rawValuesFp32.length && profile.rawValuesFp32.length % 4 === 0) {
        return {
            coordinate: profile.coordinate,
            statsType: profile.statsType,
            values: new Float32Array(profile.rawValuesFp32.slice().buffer)
        };
    }
    return {
        coordinate: profile.coordinate,
        statsType: profile.statsType,
        values: null
    };
}
export function Usage(pid): Promise<any> {
    return new Promise((resolve, reject) => {
        nodeusage.lookup(
            pid, { keepHistory: true },
            (err, info) => {
                if (err) {
                    console.log("Pid usage error: " + err);
                    reject();
                }
                resolve(info);
            }
        );
    });
}
export function Wait(time) {
    return new Promise(resolve => setTimeout(resolve, time));
}
export function EmptyTxt(file) {
    return new Promise((resolve, reject) => {
        fs.writeFile(file, "", err => {
            if (err) {
                console.log("Empty log file error: " + err);
                reject();
            }
            resolve();
        });
    });
}
export function AppendTxt(file, txt) {
    return new Promise((resolve, reject) => {
        fs.appendFile(file, JSON.stringify(txt) + "\n", err => {
            if (err) {
                console.log("Write log file error: " + err);
                reject();
            }
            resolve();
        });
    });
}
export function DiskUsage(pid) {
    return new Promise((resolve, reject) => {
        if (procfs.works) {
            procfs(pid).io((err, io) => {
                resolve(io.read_bytes);
            });
        } else {
            reject();
        }
    });
}
export function ThreadNumber(pid) {
    return new Promise((resolve, reject) => {
        if (procfs.works) {
            procfs(pid).threads((err, task) => {
                resolve(task.length);
            });
        } else {
            reject();
        }
    });
}