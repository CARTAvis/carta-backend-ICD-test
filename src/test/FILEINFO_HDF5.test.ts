import { CARTA } from "carta-protobuf";

import { Client } from "./CLIENT";
import config from "./config.json";

let testServerUrl = config.serverURL;
let testSubdirectory = config.path.QA;
let connectTimeout = config.timeout.connection;
let listFileTimeout = config.timeout.listFile;
let openFileTimeout = config.timeout.openFile;

interface AssertItem {
    registerViewer: CARTA.IRegisterViewer;
    fileInfoRequest: CARTA.IFileInfoRequest;
    fileInfoResponse: CARTA.IFileInfoResponse;
    lenComputedEntries: number;
    lenHeaderEntries: number;
};

let assertItem: AssertItem = {
    registerViewer: {
        sessionId: 0,
        clientFeatureFlags: 5,
    },
    fileInfoRequest: {
        file: "M17_SWex.hdf5",
        hdu: "",
    },
    fileInfoResponse: {
        success: true,
        message: "",
        fileInfo: {
            name: "M17_SWex.hdf5",
            type: CARTA.FileType.HDF5,
            size: 185333128,
            HDUList: ["0"],
        },
        fileInfoExtended: {
            '0': {
                dimensions: 4,
                width: 640,
                height: 800,
                depth: 25,
                stokes: 1,
                stokesVals: [],
            },
        },
    },
    lenComputedEntries: 17,
    lenHeaderEntries: 76,
};

describe("FILEINFO_HDF5: Testing if info of an HDF5 image file is correctly delivered by the backend", () => {

    let Connection: Client;
    beforeAll(async () => {
        Connection = new Client(testServerUrl);
        await Connection.open();
        await Connection.registerViewer(assertItem.registerViewer);
    }, connectTimeout);

    describe(`Go to "${testSubdirectory}" folder`, () => {
        let basePath: string;
        beforeAll(async () => {
            await Connection.send(CARTA.FileListRequest, { directory: "$BASE" });
            basePath = (await Connection.receive(CARTA.FileListResponse) as CARTA.FileListResponse).directory;
        }, listFileTimeout);

        describe(`query the info of file : ${assertItem.fileInfoRequest.file}`, () => {
            let FileInfoResponse: CARTA.FileInfoResponse;
            test(`FILE_INFO_RESPONSE should arrive within ${openFileTimeout} ms".`, async () => {
                await Connection.send(CARTA.FileInfoRequest, {
                    directory: `${basePath}/` + testSubdirectory,
                    ...assertItem.fileInfoRequest,
                });
                FileInfoResponse = await Connection.receive(CARTA.FileInfoResponse);
            }, openFileTimeout);

            test("FILE_INFO_RESPONSE.success = true", () => {
                expect(FileInfoResponse.success).toBe(true);
            });

            test(`FILE_INFO_RESPONSE.file_info.HDU_List = [${assertItem.fileInfoResponse.fileInfo.HDUList}]`, () => {
                expect(FileInfoResponse.fileInfo.HDUList[0]).toBe(assertItem.fileInfoResponse.fileInfo.HDUList[0]);
            });

            test(`FILE_INFO_RESPONSE.file_info.name = "${assertItem.fileInfoResponse.fileInfo.name}"`, () => {
                expect(FileInfoResponse.fileInfo.name).toEqual(assertItem.fileInfoResponse.fileInfo.name);
            });

            test(`FILE_INFO_RESPONSE.file_info.size = ${assertItem.fileInfoResponse.fileInfo.size}`, () => {
                expect(FileInfoResponse.fileInfo.size.toString()).toEqual(assertItem.fileInfoResponse.fileInfo.size.toString());
            });

            test(`FILE_INFO_RESPONSE.file_info.type = ${CARTA.FileType.HDF5}`, () => {
                expect(FileInfoResponse.fileInfo.type).toBe(assertItem.fileInfoResponse.fileInfo.type);
            });

            test(`FILE_INFO_RESPONSE.file_info_extended.dimensions = ${assertItem.fileInfoResponse.fileInfoExtended['0'].dimensions}`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].dimensions).toEqual(assertItem.fileInfoResponse.fileInfoExtended['0'].dimensions);
            });

            test(`FILE_INFO_RESPONSE.file_info_extended.width = ${assertItem.fileInfoResponse.fileInfoExtended['0'].width}`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].width).toEqual(assertItem.fileInfoResponse.fileInfoExtended['0'].width);
            });

            test(`FILE_INFO_RESPONSE.file_info_extended.height = ${assertItem.fileInfoResponse.fileInfoExtended['0'].height}`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].height).toEqual(assertItem.fileInfoResponse.fileInfoExtended['0'].height);
            });

            if (assertItem.fileInfoResponse.fileInfoExtended['0'].dimensions > 2) {
                test(`FILE_INFO_RESPONSE.file_info_extended.depth = ${assertItem.fileInfoResponse.fileInfoExtended['0'].depth}`, () => {
                    expect(FileInfoResponse.fileInfoExtended['0'].depth).toEqual(assertItem.fileInfoResponse.fileInfoExtended['0'].depth);
                });
            };

            if (assertItem.fileInfoResponse.fileInfoExtended['0'].dimensions > 3) {
                test(`FILE_INFO_RESPONSE.file_info_extended.stokes = ${assertItem.fileInfoResponse.fileInfoExtended['0'].stokes}`, () => {
                    expect(FileInfoResponse.fileInfoExtended['0'].stokes).toEqual(assertItem.fileInfoResponse.fileInfoExtended['0'].stokes);
                });
            };

            test(`FILE_INFO_RESPONSE.file_info_extended.stokes_vals = [${assertItem.fileInfoResponse.fileInfoExtended['0'].stokesVals}]`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].stokesVals).toEqual(assertItem.fileInfoResponse.fileInfoExtended['0'].stokesVals);
            });

            test(`len(FILE_INFO_RESPONSE.file_info_extended.computed_entries)==${assertItem.lenComputedEntries}`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].computedEntries.length).toEqual(assertItem.lenComputedEntries);
            });

            test(`assert FILE_INFO_RESPONSE.file_info_extended.computed_entries`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].computedEntries).toMatchSnapshot();
            });

            test(`len(file_info_extended.header_entries)==${assertItem.lenHeaderEntries}`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].headerEntries.length).toEqual(assertItem.lenHeaderEntries);
            });

            test(`assert FILE_INFO_RESPONSE.file_info_extended.header_entries`, () => {
                expect(FileInfoResponse.fileInfoExtended['0'].headerEntries).toMatchSnapshot();
            });
        });
    });
    afterAll(() => Connection.close());
});
