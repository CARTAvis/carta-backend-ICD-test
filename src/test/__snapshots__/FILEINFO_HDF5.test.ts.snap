// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FILEINFO_HDF5: Testing if info of an HDF5 image file is correctly delivered by the backend Go to "set_QA" folder query the info of file : M17_SWex.hdf5 assert FILE_INFO_RESPONSE.file_info_extended.computed_entries 1`] = `
Array [
  Object {
    "name": "Name",
    "value": "M17_SWex.hdf5",
  },
  Object {
    "name": "HDU",
    "value": "0",
  },
  Object {
    "name": "Shape",
    "value": "[640, 800, 25, 1]",
  },
  Object {
    "entryType": "INT",
    "name": "Number of channels",
    "numericValue": 25,
    "value": "25",
  },
  Object {
    "entryType": "INT",
    "name": "Number of stokes",
    "numericValue": 1,
    "value": "1",
  },
  Object {
    "name": "Coordinate type",
    "value": "Right Ascension, Declination",
  },
  Object {
    "name": "Projection",
    "value": "SIN",
  },
  Object {
    "name": "Image reference pixels",
    "value": "[321, 401]",
  },
  Object {
    "name": "Image reference coords",
    "value": "[18:20:20.9985, -016.12.10.0003]",
  },
  Object {
    "name": "Image ref coords (deg)",
    "value": "[275.087 deg, -16.2028 deg]",
  },
  Object {
    "name": "Pixel increment",
    "value": "-0.4\\", 0.4\\"",
  },
  Object {
    "name": "Celestial frame",
    "value": "ICRS",
  },
  Object {
    "name": "Spectral frame",
    "value": "LSRK",
  },
  Object {
    "name": "Velocity definition",
    "value": "RADIO",
  },
  Object {
    "name": "Pixel unit",
    "value": "Jy/beam",
  },
  Object {
    "name": "Restoring beam",
    "value": "2.06105\\" X 1.49126\\", -74.6267 deg",
  },
  Object {
    "name": "Has mipmaps",
    "value": "T",
  },
]
`;

exports[`FILEINFO_HDF5: Testing if info of an HDF5 image file is correctly delivered by the backend Go to "set_QA" folder query the info of file : M17_SWex.hdf5 assert FILE_INFO_RESPONSE.file_info_extended.header_entries 1`] = `
Array [
  Object {
    "name": "SCHEMA_VERSION",
    "value": "0.3",
  },
  Object {
    "name": "HDF5_CONVERTER",
    "value": "hdf_convert",
  },
  Object {
    "name": "HDF5_CONVERTER_VERSION",
    "value": "0.1.10",
  },
  Object {
    "name": "SIMPLE",
    "value": "T",
  },
  Object {
    "entryType": "INT",
    "name": "BITPIX",
    "numericValue": -32,
    "value": "-32",
  },
  Object {
    "entryType": "INT",
    "name": "NAXIS",
    "numericValue": 4,
    "value": "4",
  },
  Object {
    "entryType": "INT",
    "name": "NAXIS1",
    "numericValue": 640,
    "value": "640",
  },
  Object {
    "entryType": "INT",
    "name": "NAXIS2",
    "numericValue": 800,
    "value": "800",
  },
  Object {
    "entryType": "INT",
    "name": "NAXIS3",
    "numericValue": 25,
    "value": "25",
  },
  Object {
    "entryType": "INT",
    "name": "NAXIS4",
    "numericValue": 1,
    "value": "1",
  },
  Object {
    "name": "EXTEND",
    "value": "T",
  },
  Object {
    "entryType": "FLOAT",
    "name": "BSCALE",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "BZERO",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "BMAJ",
    "numericValue": 0.0005725136068132,
    "value": "0.0005725136068132",
  },
  Object {
    "entryType": "FLOAT",
    "name": "BMIN",
    "numericValue": 0.0004142385721207,
    "value": "0.0004142385721207",
  },
  Object {
    "entryType": "FLOAT",
    "name": "BPA",
    "numericValue": -74.62673187256,
    "value": "-74.62673187256",
  },
  Object {
    "name": "BTYPE",
    "value": "Intensity",
  },
  Object {
    "name": "OBJECT",
    "value": "M17SW",
  },
  Object {
    "comment": "/beam'",
    "name": "BUNIT",
    "value": "Jy",
  },
  Object {
    "name": "RADESYS",
    "value": "ICRS",
  },
  Object {
    "entryType": "FLOAT",
    "name": "LONPOLE",
    "numericValue": 180,
    "value": "180.0000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "LATPOLE",
    "numericValue": -16.20277777779,
    "value": "-16.20277777779",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC01_01",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC02_01",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC03_01",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC04_01",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC01_02",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC02_02",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC03_02",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC04_02",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC01_03",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC02_03",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC03_03",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC04_03",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC01_04",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC02_04",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC03_04",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PC04_04",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "name": "CTYPE1",
    "value": "RA---SIN",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRVAL1",
    "numericValue": 275.0875000001,
    "value": "275.0875000001",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CDELT1",
    "numericValue": -0.0001111111111111,
    "value": "-0.0001111111111111",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRPIX1",
    "numericValue": 321,
    "value": "321.0000000000",
  },
  Object {
    "name": "CUNIT1",
    "value": "deg",
  },
  Object {
    "name": "CTYPE2",
    "value": "DEC--SIN",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRVAL2",
    "numericValue": -16.20277777779,
    "value": "-16.20277777779",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CDELT2",
    "numericValue": 0.0001111111111111,
    "value": "0.0001111111111111",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRPIX2",
    "numericValue": 401,
    "value": "401.0000000000",
  },
  Object {
    "name": "CUNIT2",
    "value": "deg",
  },
  Object {
    "name": "CTYPE3",
    "value": "FREQ",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRVAL3",
    "numericValue": 86751396188.4,
    "value": "86751396188.40",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CDELT3",
    "numericValue": -244237.7011414,
    "value": "-244237.7011414",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRPIX3",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "name": "CUNIT3",
    "value": "Hz",
  },
  Object {
    "name": "CTYPE4",
    "value": "STOKES",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRVAL4",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CDELT4",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "CRPIX4",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "name": "CUNIT4",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PV2_1",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "PV2_2",
    "value": "0.000000000000",
  },
  Object {
    "entryType": "FLOAT",
    "name": "RESTFRQ",
    "numericValue": 86754290000,
    "value": "86754290000.00",
  },
  Object {
    "name": "SPECSYS",
    "value": "LSRK",
  },
  Object {
    "entryType": "FLOAT",
    "name": "ALTRVAL",
    "numericValue": 9999.999914138,
    "value": "9999.999914138",
  },
  Object {
    "entryType": "FLOAT",
    "name": "ALTRPIX",
    "numericValue": 1,
    "value": "1.000000000000",
  },
  Object {
    "entryType": "INT",
    "name": "VELREF",
    "numericValue": 257,
    "value": "257",
  },
  Object {
    "name": "TELESCOP",
    "value": "ALMA",
  },
  Object {
    "name": "OBSERVER",
    "value": "sishii",
  },
  Object {
    "name": "DATE-OBS",
    "value": "2016-04-03T13:02:58.800000",
  },
  Object {
    "name": "TIMESYS",
    "value": "UTC",
  },
  Object {
    "entryType": "FLOAT",
    "name": "OBSRA",
    "numericValue": 275.0875000001,
    "value": "275.0875000001",
  },
  Object {
    "entryType": "FLOAT",
    "name": "OBSDEC",
    "numericValue": -16.20277777779,
    "value": "-16.20277777779",
  },
  Object {
    "entryType": "FLOAT",
    "name": "OBSGEO-X",
    "numericValue": 2225142.180269,
    "value": "2225142.180269",
  },
  Object {
    "entryType": "FLOAT",
    "name": "OBSGEO-Y",
    "numericValue": -5440307.370349,
    "value": "-5440307.370349",
  },
  Object {
    "entryType": "FLOAT",
    "name": "OBSGEO-Z",
    "numericValue": -2481029.851874,
    "value": "-2481029.851874",
  },
  Object {
    "name": "DATE",
    "value": "2016-09-07T22:08:24.390000",
  },
  Object {
    "name": "ORIGIN",
    "value": "CASA 4.5.2-REL (r36115)",
  },
]
`;
